### PR TITLE
main: downsample in compactor, exit on complete

### PR DESCRIPTION
**Pull request by fabxc**
This adds a downsampling pass after all compaction work has been done.
By default the compactor will exit now after all compactions and
downsamplings completed, which makes deployment as a batch job easy.
The `--wait` flag restores previous behavior where the process keeps
running and waiting for new work.

### CODE REVIEW

File: cmd/thanos/compact.go [hunk 1]
```diff
@@ -136,8 +141,10 @@ func runCompact(
 
 		ctx, cancel := context.WithCancel(context.Background())
 
-		g.Add(func() error {
-			return runutil.Repeat(5*time.Minute, ctx.Done(), func() error {
+		f := func() error {
+			// Loop over bucket and compact until there's no work left.
+			didWork := false
+			for {
```
Comment (Bplotka) [inline cmd/thanos/compact.go]: just do `for didWork {` and kill 
```
if !didWork { ...
```
part?

File: cmd/thanos/compact.go [hunk 2]
```diff
@@ -151,7 +158,11 @@ func runCompact(
 				for _, g := range groups {
 					os.RemoveAll(dataDir)
 					// While we do all compactions sequentially we just compact within the top-level dir.
-					if _, err := g.Compact(ctx, dataDir, comp); err == nil {
+					id, err := g.Compact(ctx, dataDir, comp)
+					if err == nil {
+						if id != (ulid.ULID{}) {
```
Comment (Bplotka) [inline cmd/thanos/compact.go]: can we comment why we are comparing with empty ULID?

File: cmd/thanos/compact.go [hunk 3]
```diff
@@ -163,8 +174,32 @@ func runCompact(
...
+					break
+				}
+			}
+			// After all comapctions are done, work down the downsampling backlog.
+			// We run two passes of this to ensure that the 1h downsampling is generated
+			// for 5m downsamplings created in the first run.
+			level.Info(logger).Log("msg", "start first pass of downsampling")
+
+			if err := downsampleBucket(ctx, logger, bkt, dataDir); err != nil {
+				return errors.Wrap(err, "downsampling failed")
```
Comment (Bplotka) [inline cmd/thanos/compact.go]: worth to make this wrap msg unique? 

`s/downsampling failed/first phase downsampling failed`

File: cmd/thanos/compact.go (hunk 1 above)
Comment (fabxc) [reply cmd/thanos/compact.go]: We want `do ... while` semantics though and not `while`.

**Comment by fabxc** [comment]
@Bplotka addressed comments

File: cmd/thanos/compact.go (hunk 1 above)
Comment (Bplotka) [reply cmd/thanos/compact.go]: yea so change variable name to `workLeft` and put `for workLeft {` (:

File: cmd/thanos/compact.go (hunk 2 above)
Comment (Bplotka) [reply cmd/thanos/compact.go]: still not addressed

File: cmd/thanos/compact.go (hunk 3 above)
Comment (Bplotka) [reply cmd/thanos/compact.go]: not addressed

**Comment by Bplotka** [comment]
Don't see them addressed ^^

**Comment by fabxc** [comment]
What. Weird. Guess my push didn't come through.

**Comment by fabxc** [comment]
Ok, now.

**Pull request merged**